cli/global_state.rs

```diff
@@ -241,8 +241,62 @@ impl GlobalState {
   }
 }
 
+// Compilation happens if either:
+// - `checkJs` is set to true in TS config
+// - entry point is a TS file
+// - any dependency in module graph is a TS file
+fn needs_compilation(
+  compile_js: bool,
+  media_type: MediaType,
+  module_graph_files: Vec<&ModuleGraphFile>,
+) -> bool {
+  let mut needs_compilation = match media_type {
+    msg::MediaType::TypeScript | msg::MediaType::TSX | msg::MediaType::JSX => {
+      true
+    }
+    msg::MediaType::JavaScript => compile_js,
+    _ => false,
+  };
+
+  needs_compilation |= module_graph_files.iter().any(|module_file| {
+    let media_type = module_file.media_type;
+
+    media_type == (MediaType::TypeScript as i32)
+      || media_type == (MediaType::TSX as i32)
+      || media_type == (MediaType::JSX as i32)
+  });
+
+  needs_compilation
+}
+
 #[test]
 fn thread_safe() {
   fn f<S: Send + Sync>(_: S) {}
   f(GlobalState::mock(vec![]));
 }
+
+#[test]
+fn test_needs_compilation() {
+  assert!(!needs_compilation(false, MediaType::JavaScript, vec![]));
+  assert!(needs_compilation(true, MediaType::JavaScript, vec![]));
```

Comment:
Could you add one more example?
```rust
assert!(!needs_compilation(true, TODO));
```
I'm not sure what TODO should be...

Reply:
Done